### PR TITLE
refactor: Change context to be linked with view

### PR DESCRIPTION
**Pull request by tuliopereirazup**
## Description

Because of our old context implementation we are not able to create/refactor the existing ListView logic for cell reuse. 
Another reason is to remove all the differences from Android context to iOS and Web implementations.

## Related Issues

#542 

## Tests

Add new unit tests.

## Breaking Changes

BREAKING CHANGE:
- `ServerDrivenComponent.handleEvent` now expects a view that trigged the event.
- `ServerDrivenComponent.observeBindChanges` now expects the view that is owner of binding attributes.
- `Action.handleEvent` now expects origin view that is received on action's execute method.

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: android/beagle/src/main/java/br/com/zup/beagle/android/action/SendRequest.kt [hunk 1]
```diff
@@ -69,7 +69,7 @@ data class SendRequest(
     override fun execute(rootView: RootView, origin: View) {
         val viewModel = rootView.generateViewModelInstance<ActionRequestViewModel>()
 
-        viewModel.fetch(toSendRequestInternal(rootView)).observe(rootView.getLifecycleOwner(), Observer { state ->
+        viewModel.fetch(toSendRequestInternal(rootView, origin)).observe(rootView.getLifecycleOwner(), Observer { state ->
```
Comment (botbeaglezup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/action/SendRequest.kt:72]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/8d7be185c5de3cb93185332fd50fac9e603469b4/android/beagle/src/main/java/br/com/zup/beagle/android/action/SendRequest.kt#L72"></span>Line detected that is longer than the defined maximum line length in the code style.</td>
    </tr>
  </tbody>
</table>

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt [hunk 2]
```diff
@@ -16,79 +16,91 @@
 
 package br.com.zup.beagle.android.context
 
-import android.util.LruCache
+import android.view.View
 import br.com.zup.beagle.android.action.SetContextInternal
 import br.com.zup.beagle.android.jsonpath.JsonCreateTree
 import br.com.zup.beagle.android.logger.BeagleMessageLogs
+import br.com.zup.beagle.android.utils.*
```
Comment (botbeaglezup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt:23]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/8d7be185c5de3cb93185332fd50fac9e603469b4/android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt#L23"></span>br.com.zup.beagle.android.utils.* is a wildcard import. Replace it with fully qualified imports.</td>
    </tr>
  </tbody>
</table>

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt [hunk 3]
```diff
@@ -16,79 +16,91 @@
...
-        } ?: false
+    fun getContextsFromBind(originView: View, binding: Bind.Expression<*>): List<ContextData> {
+        val parentContexts = originView.getAllParentContexts()
+        val contextIds = binding.value.getExpressions().map { it.getContextId() }
+        return parentContexts.filterKeys { contextIds.contains(it) }.map { it.value.context }
     }
 
-    private fun clearContextCache(contextId: String) {
-        contexts[contextId]?.cache?.evictAll()
+    fun updateContext(view: View, setContextInternal: SetContextInternal): Boolean {
```
Comment (botbeaglezup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt:89]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/8d7be185c5de3cb93185332fd50fac9e603469b4/android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt#L89"></span>Function updateContext is nested too deeply.</td>
    </tr>
  </tbody>
</table>

File: android/beagle/src/main/java/br/com/zup/beagle/android/utils/ViewExtensions.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        if (contextBinding != null) {
+            contexts[contextBinding.context.id] = contextBinding
+        }
+        parentView = (parentView?.parent as? ViewGroup)?.getParentContextData()
+    } while (parentView != null)
+
+    return contexts
+}
+
+internal fun View.getParentContextData(): View? {
```
Comment (botbeaglezup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/utils/ViewExtensions.kt:54]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/8d7be185c5de3cb93185332fd50fac9e603469b4/android/beagle/src/main/java/br/com/zup/beagle/android/utils/ViewExtensions.kt#L54"></span>Function getParentContextData has 3 return statements which exceeds the limit of 2.</td>
    </tr>
  </tbody>
</table>

File: android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/ScreenContextViewModel.kt [hunk 5]
```diff
@@ -16,13 +16,16 @@
 
 package br.com.zup.beagle.android.view.viewmodel
 
+import android.view.View
 import androidx.lifecycle.ViewModel
 import br.com.zup.beagle.android.action.Action
 import br.com.zup.beagle.android.action.SetContextInternal
-import br.com.zup.beagle.android.context.Bind
-import br.com.zup.beagle.android.context.ContextData
+import br.com.zup.beagle.android.context.*
```
Comment (botbeaglezup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/ScreenContextViewModel.kt:23]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/8d7be185c5de3cb93185332fd50fac9e603469b4/android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/ScreenContextViewModel.kt#L23"></span>br.com.zup.beagle.android.context.* is a wildcard import. Replace it with fully qualified imports.</td>
    </tr>
  </tbody>
</table>

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/ZupIT/beagle/blob/8d7be185c5de3cb93185332fd50fac9e603469b4/android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/ScreenContextViewModel.kt#L36">android/beagle/src/main/java/br/com/zup/beagle/android/view/viewmodel/ScreenContextViewModel.kt#L36</a> - Class ‘ScreenContextViewModel’ with ‘11’ functions detected. Defined threshold inside classes is set to ‘11’</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_android">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt [hunk 6]
```diff
@@ -16,79 +16,91 @@
...
+            bind = bind
+        ))
+        viewBinding[view] = bindings
     }
 
-    fun addBindingToContext(binding: Bind.Expression<*>) {
-        binding.value.getExpressions().forEach { expression ->
-            val contextId = expression.getContextId()
-            contexts[contextId]?.bindings?.add(binding)
+    fun discoverAllContexts() {
```
Comment (matheusribeirozup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt:69]: What do you think of putting a more descriptive name?
```suggestion
    fun discoverAllContextsAndAddBindingToThem() {
```

File: android/beagle/src/test/java/br/com/zup/beagle/android/components/utils/ViewExtensionsKtTest.kt [hunk 7]
```diff
@@ -112,12 +115,17 @@ class ViewExtensionsKtTest : BaseTest() {
 
     @Test
     fun loadView_should_create_BeagleView_and_call_loadView_with_fragment() {
-        // When
+        // Given When
```
Comment (viniciusguardieirozup) [inline android/beagle/src/test/java/br/com/zup/beagle/android/components/utils/ViewExtensionsKtTest.kt:118]: I think here is just when, no?

File: android/beagle/src/main/java/br/com/zup/beagle/android/utils/ViewExtensions.kt [hunk 8]
```diff
@@ -0,0 +1,88 @@
+/*
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/utils/ViewExtensions.kt:1]: What do you think to create unit tests for these methods?

File: android/gradlew [hunk 9]
```diff
@@ -1,172 +0,0 @@
-#!/usr/bin/env sh
```
Comment (uziasferreirazup) [inline android/gradlew:1]: this file can't be deleted.

Comment (uziasferreirazup) [reply android/gradlew:1]: This file it is used to execute the task in CI and local

File: android/.gitignore [hunk 10]
```diff
@@ -0,0 +1 @@
+gradlew
```
Comment (uziasferreirazup) [inline android/.gitignore:1]: Why you put this file in android git ignore?

File: .gitignore [hunk 11]
```diff
@@ -4,6 +4,9 @@ target/
 build/
 ../out/
 .gradle/
+gradle/
+gradlew
+gradlew.bat
```
Comment (uziasferreirazup) [inline .gitignore:9]: I don't think these lines be necessary

File: android/beagle/src/main/java/br/com/zup/beagle/android/utils/ViewExtensions.kt (hunk 8 above)
Comment (Tulioh) [reply android/beagle/src/main/java/br/com/zup/beagle/android/utils/ViewExtensions.kt:1]: This methods is already testes as integrated tests

File: android/.gitignore (hunk 10 above)
Comment (Tulioh) [reply android/.gitignore]: I just did a confusion

File: android/gradlew (hunk 9 above)
Comment (Tulioh) [reply android/gradlew]: I just did a confusion

**Pull request merged**